pkg/expression/expression.go

```diff
@@ -0,0 +1,41 @@
+package expression
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/antonmedv/expr"
+)
+
+type Expression struct {
+	Comparison Comparison `yaml:"comparison" mapstructure:"comparison"`
```

Comment:
Why do we need comparison here? why not directly put the Key, Operator, and Value here?

Reply:
Expr library supports multiple kinds of [Operators](https://expr-lang.org/docs/language-definition#operators) and so the Expression is defined as an struct to support any extension. Right now we have implemented just the comparison operators.

Reply:
Even so, I don't think we need this a single struct under Expression.
We could just pass the expression as string and the expression package could just decide the logic later.

Reply:
but we are not using the package as of now. so to avoid string handling at shield, I put these as separate fields.

Reply:
also `Key` is derived from attributes. we would have to do a string handling for that as well.
Right now having separate fields keeps this simple.
WDYS?

Reply:
I was saying probably having something like this is enough, because the `Comparison` struct is not needed i believe.
```
type Expression struct {
	Key      any    `yaml:"key" mapstructure:"key"`
	Operator string `yaml:"operator" mapstructure:"operator"`
	Value    any    `yaml:"value" mapstructure:"value"`
}
```
But even if we do this, we only support binary operation though. It won't support most of the cases of the expr [Operator](https://expr-lang.org/docs/language-definition#operators).

Reply:
Some of the Operators like Regex, String(contains, startsWith, endsWith) etc can still evaluate to a boolean result.
And to support them later, we would need to introduce another struct that will have the fields to accomodate input for these.
WDYS?

Reply:
Why do we need another struct?

Why can't we just use the same Operator string, we would just need to check the operator later right? (with switch case)

Can you explain a use case?

Reply:
Also I believe the way we design the expression package should be independent of Shield's use case and flexible for expansion. This would enable it to be easily moved to salt if other services come up with similar use cases.

Reply:
I am not against that. But why do we need to have a dedicated `Comparison` struct? That is one thing that I still don't understand. I don't think it is necessary.

Reply:
Okay, so there are other class of Operators that we can add to this package, for example - Regex, String, Membership etc.

This is how is think that would look like:

```
type Expression struct {
	Comparison Comparison `yaml:"comparison" mapstructure:"comparison"`
	Regex Regex `yaml:"regex" mapstructure:"regex"`
        String String `yaml:"string" mapstructure:"string"`
        Membership Membership `yaml:"membership" mapstructure:"membership"`
}

type Comparison struct {
	Key      any    `yaml:"key" mapstructure:"key"` // "A"
	Operator string `yaml:"operator" mapstructure:"operator"` // "=="
	Value    any    `yaml:"value" mapstructure:"value"` // "B"
}

type Regex struct {
	RegexTemplate  string `yaml:"regex-template" mapstructure:"regex-template"` // "^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$"
	MatcherString string `yaml:"matcher-string" mapstructure:"matcher-string"` //  "a@b.com"
}

type String struct {
	FirstString      string    `yaml:"first-string" mapstructure:"first-string"` // "apple"
	Check string `yaml:"check" mapstructure:"check"` // "contains"
        SecondString      string    `yaml:"second-string" mapstructure:"second-string"` // "app"
}

type Membership struct {
	Member      any    `yaml:"member" mapstructure:"member"` // "name"
	Criteria string `yaml:"check" mapstructure:"check"` // "in"
        Collection      any    `yaml:"collection" mapstructure:"collection"` // {"name": "John", "age": 30}
}
```

To have compatibility for these, we need different sub strcuts for expressions.

Reply:
That is the thing.

Why cant't we just have

```
type Expression struct {
	Key      any    `yaml:"key" mapstructure:"key"` // "A"
	Operator string `yaml:"operator" mapstructure:"operator"` // "=="
	Value    any    `yaml:"value" mapstructure:"value"` // "B"
}

// Regex
type Expression struct {
	Key      any    `yaml:"key" mapstructure:"key"` // "A"
	Operator string `yaml:"operator" mapstructure:"operator"` // "^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$"
	Value    any    `yaml:"value" mapstructure:"value"`  //  "a@b.com"
}

// Regex
type Expression struct {
	Key      any    `yaml:"key" mapstructure:"key"` // "apple"
	Operator string `yaml:"operator" mapstructure:"operator"` // "contains"
	Value    any    `yaml:"value" mapstructure:"value"`  //  "app"
}

// Membership
type Expression struct {
	Key      any    `yaml:"key" mapstructure:"key"` // "name"
	Operator string `yaml:"operator" mapstructure:"operator"` // "in"
	Value    any    `yaml:"value" mapstructure:"value"`  //  {"name": "John", "age": 30}
}
```

we could just use single `key, value, operator` term to keep it simple but we support different operators.

Reply:
Sure, we can work with this as well.
Just a few concerns:
- when operator is variable it won't fit in the switch case directly (in case of regex). we would have to seperatley do a check if the operator is a regex. 
- using differnt structs gave us the flexibility to use different names for each fields. We can keep field names that are easy to understand by the user. For example, for regex, having fileds like `Regex` and `TargetString` are more intutive than having a generic struct to hold everything.
- Also let's say we have a Operator class which needs less than or more than 3 fields, then it would be confusing which field is supposed to hold what. For ex: in case of regex, how shall we decide wether `Key` or `Value` will hold the target string?
- If we still go ahead with having a struct, what shall be the name of the fields that can cover all the use cases? We would need to have very clear documentation for this.

Reply:
- The `operator` should be the defined one. It should not be a variable.
- Maintaining these structs would be harder and make the contract more stiff. I would suggest to start having a generic contract for this first.
- Now you are talking about `binary`, `ternary`, `unary` operator. We could restrict that Shield only support `binary` operator for now.
- Since the `Key` will always be fetched based on attribute, what about
```
Attribute  any    `yaml:"attribute" mapstructure:"attribute"`
Operator string `yaml:"operator" mapstructure:"operator"`
Value       any     `yaml:"value" mapstructure:"value"`
```

Reply:
Sure mas we can do this. 
One question that stays:
- in case of regex, the operator would be a regex, which can't be a constant. Every regex can be unique. How shall we handle this case? this won't fit in the Switch case statements.

Reply:
We could define regex later, could be something like `regex('xxxx')` or `match('xxxx')`
But we should avoid regex as much as possible for now because debugging regex is troublesome and it prone to inconsistent behaviour.